src/EditorFeatures/CSharp/Highlighting/KeywordHighlighters/AsyncAwaitHighlighter.cs

```diff
@@ -26,93 +33,106 @@ protected override bool IsHighlightableNode(SyntaxNode node)
                 case UsingStatementSyntax usingStatement:
-                    if (usingStatement.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword)
-                    {
-                        spans.Add(usingStatement.AwaitKeyword.Span);
-                    }
-                    break;
-
+                    TryAddAsyncOrAwaitKeyword(usingStatement.AwaitKeyword, spans);
+                    return;
                 case LocalDeclarationStatementSyntax localDeclaration:
-                    if (localDeclaration.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword && localDeclaration.UsingKeyword.Kind() == SyntaxKind.UsingKeyword)
+                    if (localDeclaration.UsingKeyword.Kind() == SyntaxKind.UsingKeyword)
                     {
-                        spans.Add(localDeclaration.AwaitKeyword.Span);
+                        TryAddAsyncOrAwaitKeyword(localDeclaration.AwaitKeyword, spans);
                     }
-                    break;
-
+                    return;
                 case CommonForEachStatementSyntax forEachStatement:
-                    if (forEachStatement.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword)
-                    {
-                        spans.Add(forEachStatement.AwaitKeyword.Span);
-                    }
-                    break;
+                    TryAddAsyncOrAwaitKeyword(forEachStatement.AwaitKeyword, spans);
+                    return;
+                case AwaitExpressionSyntax awaitExpression:
+                    TryAddAsyncOrAwaitKeyword(awaitExpression.AwaitKeyword, spans);
+                    return;
+            }
+        }
+
+        private static void TryAddAsyncModifier(SyntaxTokenList modifiers, List<TextSpan> spans)
+        {
+            foreach (var mod in modifiers)
+            {
+                if (TryAddAsyncOrAwaitKeyword(mod, spans))
+                {
+                    return;
+                }
             }
+        }
 
-            foreach (var child in node.ChildNodes())
+        private static bool TryAddAsyncOrAwaitKeyword(SyntaxToken mod, List<TextSpan> spans)
+        {
+            if (mod.IsKind(SyntaxKind.AsyncKeyword, SyntaxKind.AwaitKeyword))
             {
-                // Only recurse if we have anything to do
-                if (!child.IsReturnableConstruct())
+                // Note if there is already a highlight for the previous token, merge it with this
+                // span. That way, we highlight nested awaits with a single span.
+
+                if (spans.Count > 0)
                 {
-                    HighlightRelatedKeywords(child, spans);
+                    var previousToken = mod.GetPreviousToken();
+                    var lastSpan = spans[spans.Count - 1];
+                    if (lastSpan == previousToken.Span)
+                    {
+                        spans[spans.Count - 1] = TextSpan.FromBounds(lastSpan.Start, mod.Span.End);
+                        return true;
+                    }
                 }
```

Comment:
i preserved this logic (just in O(1) time instead of O(n)).  But, overall, i think preserving this logic is pretty unnecessary.  i don't see why we actually need it.

Reply:
It highlights the entire async code section rather than the highlighting the nested inside keywords `Async` and `Await`.
Personally i prefer the latter.